cpp/src/parquet/chunker_internal.cc

```diff
@@ -0,0 +1,250 @@
+                                                          int64_t num_levels,
+                                                          const T& leaf_array) {
+  std::vector<Chunk> chunks;
+  bool has_def_levels = level_info_.def_level > 0;
+  bool has_rep_levels = level_info_.rep_level > 0;
+
+  if (!has_rep_levels && !has_def_levels) {
+    // fastest path for non-nested non-null data
+    int64_t offset = 0;
+    int64_t prev_offset = 0;
+    while (offset < num_levels) {
+      Roll(leaf_array.GetView(offset));
+      ++offset;
+      if (NeedNewChunk()) {
+        chunks.emplace_back(prev_offset, prev_offset, offset - prev_offset);
+        prev_offset = offset;
+      }
+    }
+    if (prev_offset < num_levels) {
+      chunks.emplace_back(prev_offset, prev_offset, num_levels - prev_offset);
+    }
+  } else if (!has_rep_levels) {
+    // non-nested data with nulls
+    int64_t offset = 0;
+    int64_t prev_offset = 0;
+    while (offset < num_levels) {
+      Roll(def_levels[offset]);
+      Roll(leaf_array.GetView(offset));
+      ++offset;
+      if (NeedNewChunk()) {
+        chunks.emplace_back(prev_offset, prev_offset, offset - prev_offset);
+        prev_offset = offset;
+      }
+    }
+    if (prev_offset < num_levels) {
+      chunks.emplace_back(prev_offset, prev_offset, num_levels - prev_offset);
+    }
+  } else {
+    // nested data with nulls
+    bool has_leaf_value;
+    bool is_record_boundary;
+    int16_t def_level;
+    int16_t rep_level;
+    int64_t value_offset = 0;
+    int64_t record_level_offset = 0;
+    int64_t record_value_offset = 0;
+
+    for (int64_t level_offset = 0; level_offset < num_levels; ++level_offset) {
+      def_level = def_levels[level_offset];
+      rep_level = rep_levels[level_offset];
+
+      has_leaf_value = def_level >= level_info_.repeated_ancestor_def_level;
+      is_record_boundary = rep_level == 0;
+
+      Roll(def_level);
+      Roll(rep_level);
+      if (has_leaf_value) {
+        Roll(leaf_array.GetView(value_offset));
+      }
+
+      if (is_record_boundary && NeedNewChunk()) {
+        auto levels_to_write = level_offset - record_level_offset;
+        if (levels_to_write > 0) {
+          chunks.emplace_back(record_level_offset, record_value_offset, levels_to_write);
+          record_level_offset = level_offset;
+          record_value_offset = value_offset;
+        }
+      }
+
+      if (has_leaf_value) {
+        ++value_offset;
+      }
+    }
+
+    auto levels_to_write = num_levels - record_level_offset;
+    if (levels_to_write > 0) {
+      chunks.emplace_back(record_level_offset, record_value_offset, levels_to_write);
+    }
+  }
+
+  return chunks;
+}
+
+#define PRIMITIVE_CASE(TYPE_ID, ArrowType)               \
+  case ::arrow::Type::TYPE_ID:                           \
+    return Calculate(def_levels, rep_levels, num_levels, \
+                     static_cast<const ::arrow::ArrowType##Array&>(values));
+
```

Comment:
Generating a separate `Calculate` for each primitive type is going to be wasteful. Some implementations should be shared, for example all 64-bit numeric types (int64, uint64, float64, timestamp, etc.). Same for string/binary, large_string/large_binary...

Reply:
I updated the implementation to use the same code path for same width types, also switched to use the raw buffer pointers. I introduced a `RollValue` closure to handle different paths. Please take a look again.

Reply:
This is fixed.